lib/ide/src/double_representation/graph.rs

```diff
@@ -68,27 +99,84 @@ pub fn traverse_for_definition
-    /// inputs and outputs).
-    pub nodes:Vec<node::NodeInfo>,
 }
 
 impl GraphInfo {
     /// Describe graph of the given definition.
     pub fn from_definition(source:DefinitionInfo) -> GraphInfo {
-        let nodes = Self::from_function_binding(source.ast.clone());
-        GraphInfo {source,nodes}
+        GraphInfo {source}
     }
 
     /// Lists nodes in the given binding's ast (infix expression).
-    fn from_function_binding(ast:known::Infix) -> Vec<node::NodeInfo> {
+    fn from_function_binding(ast:known::Infix) -> Vec<NodeInfo> {
         let body = ast.rarg.clone();
         if let Ok(body_block) = known::Block::try_new(body.clone()) {
             block_nodes(&body_block)
         } else {
             expression_node(body)
         }
     }
+
+    /// Gets all known nodes in this graph (does not include special pseudo-nodes like graph
+    /// inputs and outputs).
+    pub fn nodes(&self) -> Vec<NodeInfo> {
+        Self::from_function_binding(self.source.ast.clone())
+    }
+
+    /// Adds a new node to this graph.
+    pub fn add_node
+    (&mut self, line_ast:Ast, location_hint:LocationHint) -> FallibleResult<ID> {
+        let block  = self.source.ast.rarg.clone();
+        let block  = known::Block::try_from(block)?;
+
+        let first_line = block.first_line.clone();
+        let mut lines  = block.lines.clone();
+        lines.insert(0, ast::BlockLine{elem:Some(first_line.elem),off:first_line.off});
+
+        let find_position = |id| {
+            lines.iter().find_position(|line| {
+                line.elem.as_ref().map(|line_ast| {
+                    NodeInfo::from_line_ast(line_ast).map(|node| node.id() == id).unwrap_or(false)
+                }).unwrap_or(false)
+            }).map(|(index,_)| index).ok_or(IDNotFound{id})
+        };
+
+        let index = match location_hint {
+            LocationHint::Start      => 0,
+            LocationHint::End        => lines.len(),
+            LocationHint::After(id)  => find_position(id)? + 1,
+            LocationHint::Before(id) => find_position(id)?
+        };
+
+        lines.insert(index, ast::BlockLine { elem: Some(line_ast), off: 0 });
+        let first_line = lines.remove(0);
+        let first_line = ast::BlockLine{elem:first_line.elem.unwrap(),off:first_line.off};
+
+        let ty          = block.ty.clone();
+        let indent      = block.indent.clone();
+        let empty_lines = block.empty_lines.clone();
+        let is_orphan   = block.is_orphan.clone();
+        let block = Ast::new(ast::Block {ty,indent,empty_lines,first_line,lines,is_orphan}, None);
+
+        let larg  = self.source.ast.larg.clone();
+        let loff  = self.source.ast.loff.clone();
+        let opr   = self.source.ast.opr.clone();
+        let roff  = self.source.ast.roff.clone();
+        let rarg  = block.try_into()?;
+        let infix = Ast::new(ast::Infix {larg,loff,opr,roff,rarg}, None);
```

Comment:
add these to `impl<T> KnownAst<T>`

```
    pub fn new_unchecked(ast:Ast) -> KnownAst<T> {
        KnownAst {ast,phantom:default()}
    }

    pub fn new(shape:T, id:Option<Id>) -> KnownAst<T> {
        let ast = Ast::new(shape,id);
        Self::new_unchecked(ast)
    }
```

Then you should be able to simplify this — just create known Ast directly from ast::Infix, with no intermediate Ast